libp2p/builders.nim

```diff
@@ -0,0 +1,198 @@
+import
+  options, tables, chronos, bearssl,
+  switch, peerid, peerinfo, stream/connection, multiaddress,
+  crypto/crypto, transports/[transport, tcptransport],
+  muxers/[muxer, mplex/mplex],
+  protocols/[identify, secure/secure, secure/noise],
+  connmanager, upgrademngrs/muxedupgrade
+
+export
+  switch, peerid, peerinfo, connection, multiaddress, crypto
+
+type
+  SecureProtocol* {.pure.} = enum
+    Noise,
+    Secio {.deprecated.}
+
+  TcpTransportOpts = object
+    enable: bool
+    flags: set[ServerFlags]
+
+  MplexOpts = object
+    enable: bool
+    newMuxer: MuxerConstructor
+
+  SwitchBuilder* = ref object
+    privKey: Option[PrivateKey]
+    address: MultiAddress
+    secureManagers: seq[SecureProtocol]
+    mplexOpts: MplexOpts
+    tcpTransportOpts: TcpTransportOpts
+    rng: ref BrHmacDrbgContext
+    inTimeout: Duration
+    outTimeout: Duration
+    maxConnections: int
+    maxIn: int
+    maxOut: int
+    maxConnsPerPeer: int
+    protoVersion: string
+    agentVersion: string
+
+proc init*(T: type[SwitchBuilder]): T =
+  SwitchBuilder(
+    privKey: none(PrivateKey),
+    address: MultiAddress.init("/ip4/127.0.0.1/tcp/0").tryGet(),
+    secureManagers: @[],
+    tcpTransportOpts: TcpTransportOpts(),
+    rng: newRng(),
```

Comment:
would be better to not create resources like this eagerly, they might not be available on all systems and are a cost at creation time

Reply:
You mean `newRng()` - good point in any case.